Feature/workflow bellman values

h-fournier opened the pull request.

h-fournier committed:
Merge branch 'develop' into feature/workflow_bellman_values

h-fournier committed:
Small fixes

h-fournier committed:
Cleaner version of GridEvaluator using GridCollection

h-fournier committed:
Update after cleaning

h-fournier committed:
Merge commit 'bbdd14ed5d124ead9916d6db0d0942bb4733ef82' into feature/workflow_bellman_values

h-fournier committed:
Create new class for the generation and modification of mps for Bellman Value calculations

h-fournier committed:
Use gridEvaluator for a GridDefinition and not GridCollection

h-fournier committed:
Antares Solver is called ones to solve all problems and then problems are used and modified for the BellmanValues calculation

h-fournier committed:
Compute bellman values (simplified version with interpolations)

tbittar commented on src/cpp/exe/full_grid_evaluation/main.cpp:
```diff
@@ -0,0 +1,107 @@
+    Output::ConcurrentInsertionMap<Antares::Solver::WeeklyProblemId, std::shared_ptr<Problem>>
+      problems;
+    Antares::Solver::Optimization::OptimizationOptions optOptions;
+    optOptions.linearSolver = "coin";
```
Ce serait quand même sympa d'avoir une petite option pour utiliser Xpress, ce sera notre cas nominal

h-fournier replied:
Ok, fait

tbittar commented on src/cpp/grid_evaluator/GridEvaluator.cpp:
```diff
@@ -20,23 +21,26 @@ std::atomic<double> totalPbModifTimer = 0; ///< Total time spent modifying subpr
-void GridEvaluator::setThreads(int nbThreads)
-{
-    this->nbThreads = nbThreads;
+    this->logger = std::move(logger);
```
You cannot use the syntax `GridEvaluator(...): logger(std::move(logger))` ? (Used here for instance https://github.com/AntaresSimulatorTeam/antares-xpansion/blob/fc808f2734d7a1fd514409465fcbbc44e9922a27/src/cpp/benders/benders_core/Worker.cpp#L197)

h-fournier replied:
Yes possible

tbittar commented on src/cpp/grid_evaluator/ReservoirManagement.cpp:
This whole class is made to read specific input data from the antares study. I wonder if we need to do this in C++, as the format of antares study may change in the future and then this part of code will need to be adapted each time. Do you think it is possible to have a part in Python (using antaresCraft package), that is responsible for reading the relevant data to the study and providing this as input for the `GridEvaluator` ? We need to think of this with the whole workflow

tbittar commented on src/cpp/multisolver_interface/include/antares-xpansion/multisolver_interface/SolverAbstract.h:
```diff
@@ -197,7 +197,7 @@ enum SOLVER_STATUS
  * \brief Virtual class to implement solvers methods
  */
-class SolverAbstract
+class SolverAbstract: public std::enable_shared_from_this<SolverAbstract>
```
Why do you need this ?

h-fournier replied:
We need to clone Problem instances, which inherit from SolverAbstract. This is necessary because ProblemGenerationForWaterValueCalculation modifies each Problem multiple times — once per gridElement in the gridDefinition.

Using std::enable_shared_from_this<SolverAbstract> allows us to safely create shared clones of problems via shared_from_this(). While it might be possible to replace this with std::unique_ptr to avoid inheriting from enable_shared_from_this, doing so would require implementing a proper copy constructor for SolverAbstract, which is currently deleted. So for now, this approach is the most practical.

h-fournier committed:
Fix missing include

h-fournier committed:
Add test for bellman values computation and work in progress for bellman values outputs

h-fournier committed:
Fix for the test to pass, not clean, a solution needs to be though about

h-fournier committed:
test solve path issue on centos during tests

h-fournier committed:
test solve  issue on centos during tests

h-fournier committed:
test solve issue on centos during tests

h-fournier committed:
Compute Bellman values from startWeek to endWeek with default values

h-fournier committed:
Refactoring : Remove Bellman value computation from gridEvaluator

h-fournier committed:
Add BellmanValuesExeOptions. The options of the exe are : studyPath, solverName and nbThreads

h-fournier committed:
fix initialization

h-fournier committed:
Add default values for options

h-fournier committed:
Initialize documentation on bellman_values.exe usage

h-fournier committed:
Forgot to stage the main

h-fournier committed:
Add unit test for bellman values calculation using a mock GridEvaluator.
Add options for the bellman_value_exe for the start and end week and number of stock levels for the belleman value calculation.

h-fournier committed:
Add antares-format output option and update documentation

h-fournier committed:
Use same solver for the bellman value workflow and add start and end week for gridEvaluator

h-fournier committed:
Small fix

h-fournier committed:
Compute Water values and save them

h-fournier committed:
Compute water valeus from belleman values and levels.
Get min and max HydroPower (max_turbining and max_pumping) from the input study.
Reset the bound of HydProd variables because it might have been modified by the heuristic.

h-fournier committed:
Modify the main with the new structure

h-fournier committed:
Merge branch 'develop' into feature/workflow_bellman_values

h-fournier committed:
Add reading of data for rule curves and new interpolator class for the linear interpolation

h-fournier committed:
Rename one test

h-fournier committed:
Implementation of the rule curves penalties

h-fournier committed:
Generate refs from Python and implement test for bellman_values computation with penalties.
The test is not passing for the moment as the values of week 1 and 2 are not exactly the same (other 50 weeks values are good)

h-fournier committed:
Add Bellman values tests to the CI

h-fournier committed:
Update ref data with good reference study

h-fournier committed:
Add unit test with penalties and modify the refs to be closer to a real life example

h-fournier committed:
New way of adding penalties (same as Python)

h-fournier committed:
Rename rewards to costs

h-fournier committed:
Add new tests with force_final_level and fix indices issues

Hugo-Frnr committed:
New parameters for H2 experiment

h-fournier committed:
Fix save water values using antares format (begin at week 1 instead of 0)

Hugo-Frnr committed:
Fix issue with scenario indices

h-fournier committed:
Fix failing tests

h-fournier committed:
Support reading windows file while parsing hydro.ini

h-fournier committed:
Fix index error

h-fournier committed:
Fix endWeek issue.
Now if the end or starting week indicated in the command line are outside of the bounds of the problem, we take limit the starting and ending week to the available ones

PAugVallin committed:
Merge commit '721742ac5a440207047619a33af1b3ab8180de42' into feature/workflow_bellman_values

PAugVallin committed:
fix: errors due to hasty merge in the previous commit

h-fournier committed:
Merge commit 'a09e87a3a7a44b1f332420501647bcd6183cc10a' into feature/workflow_bellman_values

h-fournier committed:
Do all the workflow in memory, avoid saving and reading MPS

PAugVallin committed:
chore: added a "write-pb-files" flag to water_values_exe, user may now choose to have the problem files saved to disk (default: false)

PAugVallin committed:
Updated documentation to account for --write-pb-files flag

PAugVallin committed:
Added flag problem-format for writing solved
problems to MPS or optimized file format + renamed
flag to keepMps

PAugVallin committed:
fix: using \t as Antares-format separator in the output water values file + padding first line with 0s

PAugVallin committed:
Added possibility of reading a penalties.yaml file at the root of a study to configure a simulation

PAugVallin committed:
Improved: parsing YAML for penalties

PAugVallin committed:
fix: missing a couple of stl includes

PAugVallin committed:
updated documentation for water values compuation

PAugVallin committed:
Improved logging system in water values computation

PAugVallin committed:
Checking for correct "use heuristic" and "use water" values in hydro.ini

PAugVallin committed:
fix: bad exception handling in ReservoirManagement

PAugVallin committed:
Removing use of <format> STL, not available on all compilers

h-fournier committed:
Fix the grid generation if min and max values are similar + add some safety when reading the gridElements

h-fournier committed:
Fix checks for valid gridElement

h-fournier committed:
Fix compilation error

h-fournier committed:
Add a 0 value to the end of each line for the water_values using AntaresFormat

h-fournier committed:
Merge branch 'develop' into feature/workflow_bellman_values

h-fournier committed:
Revert data that should not have been added

h-fournier committed:
PR review

h-fournier committed:
Fix format

h-fournier committed:
Fix loop using tbb (the std::execution::par was badly supported by gcc)

h-fournier committed:
Fix  logs
